File: generate_playlist.py
import os
import json
from datetime import datetime
import re

def load_config():
    """Загрузка конфигурации из config.json"""
    try:
        with open('config.json', 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        print("❌ Файл config.json не найден")
        return None
    except json.JSONDecodeError:
        print("❌ Ошибка в формате config.json")
        return None

def find_album_cover(album_path, config):
    """Поиск обложки альбома"""
    for cover_name in config['coverNames']:
        for ext in config['supportedImages']:
            cover_file = f"{cover_name}{ext}"
            cover_path = os.path.join(album_path, cover_file)
            if os.path.exists(cover_path):
                return cover_path
    return None

def find_track_images(track_files, album_path, config):
    """Поиск индивидуальных обложек для треков"""
    track_images = {}
    for track_file in track_files:
        track_name = os.path.splitext(track_file)[0]
        for ext in config['supportedImages']:
            image_file = f"{track_name}{ext}"
            image_path = os.path.join(album_path, image_file)
            if os.path.exists(image_path):
                track_images[track_file] = image_path
                break
    return track_images

def scan_albums(config):
    """Сканирование папки с альбомами"""
    albums_folder = config['albumsFolder']
    if not os.path.exists(albums_folder):
        print(f"❌ Папка {albums_folder} не найдена")
        return []

    albums = []
    
    for album_dir in os.listdir(albums_folder):
        album_path = os.path.join(albums_folder, album_dir)
        
        # Пропускаем, если не директория
        if not os.path.isdir(album_path):
            continue
            
        # Поиск аудиофайлов
        audio_files = []
        for file in os.listdir(album_path):
            file_ext = os.path.splitext(file)[1].lower()
            if file_ext in config['supportedAudio']:
                audio_files.append(file)
                
        if not audio_files:
            print(f"⚠️  В альбоме '{album_dir}' не найдено аудиофайлов")
            continue
            
        # Поиск обложки альбома
        cover_path = find_album_cover(album_path, config)
        if not cover_path:
            print(f"⚠️  У альбома '{album_dir}' отсутствует обложка")
            
        # Сортировка аудиофайлов для более предсказуемого порядка
        audio_files.sort()
        
        album_data = {
            'id': re.sub(r'[^a-zA-Z0-9_-]', '-', album_dir),
            'title': album_dir,
            'path': f"{albums_folder}/{album_dir}",
            'cover': cover_path.replace('\\', '/') if cover_path else None,
            'tracksFile': f"data/albums/{album_dir}.js",
            'trackCount': len(audio_files)
        }
        
        albums.append(album_data)
        print(f"✅ Найден альбом '{album_dir}' с {len(audio_files)} треками")
        
    return albums

def generate_album_tracks(album_path, album_title, config):
    """Генерация JS-файла с треками альбома"""
    audio_files = []
    for file in os.listdir(album_path):
        file_ext = os.path.splitext(file)[1].lower()
        if file_ext in config['supportedAudio']:
            audio_files.append(file)
    
    # Сортировка для предсказуемого порядка
    audio_files.sort()
    
    # Поиск индивидуальных обложек для треков
    track_images = find_track_images(audio_files, album_path, config)
    
    tracks = []
    for audio_file in audio_files:
        track_name = os.path.splitext(audio_file)[0]
        # Улучшенная обработка названия трека (убираем нумерацию в начале)
        clean_title = re.sub(r'^\d{1,3}[ _\-\.]*', '', track_name).strip()
        if not clean_title:
            clean_title = track_name
        
        track_data = {
            'title': clean_title,
            'file': f"{album_path}/{audio_file}".replace('\\', '/'),
            'cover': track_images.get(audio_file),
            'artist': config['defaultArtist']
        }
        tracks.append(track_data)
    
    # Создание папки data/albums если не существует
    output_dir = os.path.join(config['outputFolder'], 'albums')
    os.makedirs(output_dir, exist_ok=True)
    
    # Генерация JS файла
    output_file = os.path.join(output_dir, f"{os.path.basename(album_path)}.js")
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(f"// Авто-генерация: {timestamp}\n")
        f.write(f"// Альбом: {album_title}\n")
        f.write("// Не редактируйте вручную — изменения будут перезаписаны\n\n")
        f.write("const ALBUM_TRACKS = ")
        json.dump(tracks, f, ensure_ascii=False, indent=2)
        f.write(";\n")
    
    return output_file

def main():
    print("🎵 Запуск генератора плейлиста...")
    
    config = load_config()
    if not config:
        return
    
    # Сканирование альбомов
    albums = scan_albums(config)
    
    if not albums:
        print("❌ Альбомы не найдены")
        return
    
    # Создание папки data если не существует
    os.makedirs(config['outputFolder'], exist_ok=True)
    
    # Генерация JS файла для каждого альбома
    for album in albums:
        album_path = os.path.join(config['albumsFolder'], album['title'])
        try:
            track_file = generate_album_tracks(album_path, album['title'], config)
            print(f"✅ Сгенерирован файл треков: {track_file}")
        except Exception as e:
            print(f"❌ Ошибка при генерации треков для альбома {album['title']}: {str(e)}")
    
    # Генерация общего файла albums.js
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    albums_js_path = os.path.join(config['outputFolder'], 'albums.js')
    
    with open(albums_js_path, 'w', encoding='utf-8') as f:
        f.write(f"// Авто-генерация: {timestamp}\n")
        f.write("// Не редактируйте вручную — изменения будут перезаписаны\n\n")
        f.write("const ALBUMS = ")
        json.dump(albums, f, ensure_ascii=False, indent=2)
        f.write(";\n")
    
    print(f"✅ Сгенерирован основной файл: {albums_js_path}")
    
    # Итоговая статистика
    total_tracks = sum(album['trackCount'] for album in albums)
    print(f"\n📊 Итоговая статистика:")
    print(f"   Альбомов: {len(albums)}")
    print(f"   Всего треков: {total_tracks}")

if __name__ == "__main__":
    main()

File: generate-playlist.js
/**
 * DAJET Playlist Generator
 * Generates JavaScript files with album and track information
 */

const fs = require('fs');
const path = require('path');

// Default configuration
const defaultConfig = {
  albumsFolder: 'albums',
  outputFolder: 'data',
  defaultArtist: 'BEST',
  supportedAudio: ['.mp3', '.wav', '.flac', '.ogg', '.m4a'],
  supportedImages: ['.jpg', '.jpeg', '.png', '.webp'],
  coverNames: ['cover', 'folder', 'album'],
  sortAlbums: 'name',
  sortTracks: 'name'
};

// Load configuration
function loadConfig() {
  const configPath = './config.json';
  if (fs.existsSync(configPath)) {
    const configContent = fs.readFileSync(configPath, 'utf8');
    return { ...defaultConfig, ...JSON.parse(configContent) };
  }
  return defaultConfig;
}

// Get current timestamp
function getTimestamp() {
  const now = new Date();
  return now.toISOString().replace('T', ' ').substring(0, 19);
}

// Find cover image in a folder
function findCover(folderPath, config) {
  for (const coverName of config.coverNames) {
    for (const ext of config.supportedImages) {
      const coverPath = path.join(folderPath, coverName + ext);
      if (fs.existsSync(coverPath)) {
        return path.relative('.', coverPath).replace(/\\/g, '/');
      }
    }
  }
  
  // If no standard cover found, look for any image file
  try {
    const files = fs.readdirSync(folderPath);
    for (const file of files) {
      const ext = path.extname(file).toLowerCase();
      if (config.supportedImages.includes(ext)) {
        return path.relative('.', path.join(folderPath, file)).replace(/\\/g, '/');
      }
    }
  } catch (error) {
    console.error(`Error reading folder ${folderPath}:`, error.message);
  }
  
  return null;
}

// Get all audio files in a folder
function getAudioFiles(folderPath, config) {
  const files = [];
  
  try {
    const dirEntries = fs.readdirSync(folderPath);
    
    for (const entry of dirEntries) {
      const fullPath = path.join(folderPath, entry);
      const stat = fs.statSync(fullPath);
      
      if (stat.isFile()) {
        const ext = path.extname(entry).toLowerCase();
        if (config.supportedAudio.includes(ext)) {
          const fileNameWithoutExt = path.basename(entry, path.extname(entry));
          
          // Look for individual track cover
          let trackCover = null;
          for (const imgExt of config.supportedImages) {
            const possibleCoverPath = path.join(folderPath, fileNameWithoutExt + imgExt);
            if (fs.existsSync(possibleCoverPath)) {
              trackCover = path.relative('.', possibleCoverPath).replace(/\\/g, '/');
              break;
            }
          }
          
          files.push({
            file: path.relative('.', fullPath).replace(/\\/g, '/'),
            cover: trackCover,
            title: fileNameWithoutExt,
            artist: config.defaultArtist
          });
        }
      }
    }
  } catch (error) {
    console.error(`Error reading folder ${folderPath}:`, error.message);
  }
  
  // Sort tracks
  if (config.sortTracks === 'name') {
    files.sort((a, b) => a.title.localeCompare(b.title));
  }
  
  return files;
}

// Process a single album folder
function processAlbum(albumFolderPath, config) {
  const albumName = path.basename(albumFolderPath);
  
  // Find album cover
  const cover = findCover(albumFolderPath, config);
  
  // Get all audio files
  const tracks = getAudioFiles(albumFolderPath, config);
  
  // Generate album ID (URL-safe)
  const albumId = albumName
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, '')
    .replace(/\s+/g, '-')
    .trim();
  
  // Try to extract year from folder name (4 digits between 1900 and 2100)
  const yearMatch = albumName.match(/\b(19\d{2}|20\d{2})\b/);
  const year = yearMatch ? parseInt(yearMatch[0], 10) : null;
  
  return {
    id: albumId,
    title: albumName,
    path: path.relative('.', albumFolderPath).replace(/\\/g, '/'),
    cover: cover,
    tracksFile: `data/albums/${encodeURIComponent(albumName)}.js`,
    trackCount: tracks.length,
    year: year
  };
}

// Write album tracks to a file
function writeAlbumTracks(album, tracks, config) {
  const outputPath = path.join(config.outputFolder, 'albums', encodeURIComponent(album.title) + '.js');
  
  // Create output directory if it doesn't exist
  const outputDir = path.dirname(outputPath);
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }
  
  const content = `// Авто-генерация: ${getTimestamp()}
// Альбом: ${album.title}
// Не редактируйте вручную — изменения будут перезаписаны

const ALBUM_TRACKS = ${JSON.stringify(tracks, null, 2)};
`;
  
  fs.writeFileSync(outputPath, content);
  console.log(`✅ Записан файл: ${outputPath} (${tracks.length} треков)`);
}

// Main function to generate playlists
function generatePlaylist() {
  const config = loadConfig();
  console.log(`🎵 Запуск генерации плейлистов...`);
  console.log(`📁 Папка с альбомами: ${config.albumsFolder}`);
  console.log(`📂 Папка вывода: ${config.outputFolder}`);
  console.log('');
  
  // Create output directory structure
  if (!fs.existsSync(config.outputFolder)) {
    fs.mkdirSync(config.outputFolder, { recursive: true });
  }
  
  const albumsOutputDir = path.join(config.outputFolder, 'albums');
  if (!fs.existsSync(albumsOutputDir)) {
    fs.mkdirSync(albumsOutputDir, { recursive: true });
  }
  
  // Read all folders in albums directory
  if (!fs.existsSync(config.albumsFolder)) {
    console.error(`❌ Папка ${config.albumsFolder} не найдена!`);
    return;
  }
  
  const albumFolders = fs.readdirSync(config.albumsFolder)
    .map(folder => path.join(config.albumsFolder, folder))
    .filter(filePath => fs.statSync(filePath).isDirectory());
  
  const albums = [];
  let totalTracks = 0;
  
  for (const albumFolder of albumFolders) {
    console.log(`🔍 Обработка альбома: ${path.basename(albumFolder)}`);
    
    const album = processAlbum(albumFolder, config);
    const tracks = getAudioFiles(albumFolder, config);
    
    if (tracks.length > 0) {
      albums.push(album);
      totalTracks += tracks.length;
      
      // Write tracks file
      writeAlbumTracks(album, tracks, config);
    } else {
      console.log(`⚠️  В альбоме ${path.basename(albumFolder)} не найдено аудиофайлов`);
    }
  }
  
  // Sort albums
  if (config.sortAlbums === 'name') {
    albums.sort((a, b) => a.title.localeCompare(b.title));
  } else if (config.sortAlbums === 'year' && albums.every(album => album.year !== null)) {
    albums.sort((a, b) => (a.year || 0) - (b.year || 0));
  }
  
  // Write main albums file
  const albumsFilePath = path.join(config.outputFolder, 'albums.js');
  const albumsContent = `// Авто-генерация: ${getTimestamp()}
// Не редактируйте вручную — изменения будут перезаписаны

const ALBUMS = ${JSON.stringify(albums, null, 2)};
`;
  
  fs.writeFileSync(albumsFilePath, albumsContent);
  console.log(`✅ Записан основной файл: ${albumsFilePath}`);
  
  // Calculate total data size
  let totalSize = 0;
  const allFiles = [albumsFilePath, ...albums.map(album => 
    path.join(config.outputFolder, 'albums', encodeURIComponent(album.title) + '.js')
  )];
  
  for (const filePath of allFiles) {
    if (fs.existsSync(filePath)) {
      const stats = fs.statSync(filePath);
      totalSize += stats.size;
    }
  }
  
  console.log('');
  console.log('📊 Результаты:');
  console.log(`✅ Найдено альбомов: ${albums.length}`);
  console.log(`✅ Всего треков: ${totalTracks}`);
  console.log(`✅ Сгенерировано файлов: ${allFiles.length}`);
  console.log(`📊 Размер данных: ${(totalSize / 1024).toFixed(1)} КБ`);
  console.log('');
  console.log('✨ Готово!');
}

// Run the generator
if (require.main === module) {
  generatePlaylist();
}

module.exports = {
  generatePlaylist,
  loadConfig,
  processAlbum,
  getAudioFiles
};